### PR TITLE
WDP-85

### PR DESCRIPTION
**Pull request by lukaszcwiklinski**
https://kindergarten.stxnext.pl/browse/WDP-85

### CODE REVIEW

File: src/presence_analyzer/utils.py [hunk 1]
```diff
@@ -15,8 +17,42 @@
...
 
 log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+lock = threading.Lock()
 
 
+def cache(expire_time=60):
+    """
+    Cache decorator. Return cached data if it's not expired.
+    """
+    cache = {}
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: hold cache and timeout in one dict

File: src/presence_analyzer/utils.py [hunk 2]
```diff
@@ -15,8 +17,45 @@
...
+        def cache_wrapper(*args, **kwargs):
+            if kwargs:
+                key = args, frozenset(kwargs.items())
+            else:
+                key = args
+            now = time.time()
+
+            with lock:
+                if key in cache:
+                    if cache[key]['timeout'] > now:
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: merge this if with if in line 42

File: src/presence_analyzer/utils.py [hunk 3]
```diff
@@ -15,8 +17,45 @@
...
+                        return cache[key]['value']
+                else:
+                    timeout = now + expire_time
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: you don't have to assign this value if u use it once

File: src/presence_analyzer/utils.py [hunk 4]
```diff
@@ -15,8 +17,45 @@
...
+
+    def decorator_wrapper(function):
+        lock = threading.Lock()
+
+        def cache_wrapper(*args, **kwargs):
+            if kwargs:
+                key = args, frozenset(kwargs.items())
+            else:
+                key = args
+            now = time.time()
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: explain code 34-39 line please, why you doing this

Comment (lukaszcwiklinski) [inline src/presence_analyzer/utils.py]: I could use just "key = args, frozenset(kwargs.items())", but I thought it's more secure and optimized solution.

File: src/presence_analyzer/tests.py [hunk 5]
```diff
@@ -242,6 +242,37 @@ def tearDown(self):
...
+        self.assertEqual(f(4), 4)
+        self.assertItemsEqual(
+            utils.cache['first_key'].keys(),
+            ['value', 'timeout']
+        )
+        self.assertEqual(
+            utils.cache['first_key']['timeout'],
+            1485955538.213812
+        )
+        self.assertTrue(utils.cache, value_timeout)
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: Bad assert

File: src/presence_analyzer/tests.py [hunk 6]
```diff
@@ -242,6 +243,71 @@ def tearDown(self):
         """
         pass
 
+    def test_cache(self):
+        """
+        Test caching decorator.
+        """
+        utils.memoize(60)(utils.get_data())
```
Comment (piotrgrundas) [inline src/presence_analyzer/tests.py]: `utils.memoize(60)` is not needed since 
image

File: src/presence_analyzer/utils.py [hunk 7]
```diff
@@ -15,7 +18,38 @@
...
+                    return cache[key]['value']
+                else:
+                    result = function(*args, **kwargs)
+                    cache[key] = {
+                        'value': result,
+                        'timeout': now + expire_time
+                    }
+
+                    return result
+
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: Without empty lines

File: src/presence_analyzer/utils.py [hunk 8]
```diff
@@ -5,6 +5,9 @@
 
 import csv
 import logging
+import threading
+import pickle
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: sort imports

**Pull request merged**